lib/buffer-writer.js

```diff
@@ -0,0 +1,287 @@
+You can use close({ resize: true }) to resize header`)
+  }
+}
+
+/**
+ *
+ * @param {CarBufferWriter} writer
+ * @param {number} byteLength
+ */
+export const resizeHeader = (writer, byteLength) => {
+  const { bytes, headerSize } = writer
+  // Move data section to a new offset
+  bytes.set(bytes.subarray(headerSize, writer.byteOffset), byteLength)
+  // Update header size & byteOffset
+  writer.byteOffset += byteLength - headerSize
+  writer.headerSize = byteLength
+}
+
+/**
+ *
+ * @param {CarBufferWriter} writer
+ * @param {number[]|Uint8Array} bytes
+ */
+
+const writeBytes = (writer, bytes) => {
+  writer.bytes.set(bytes, writer.byteOffset)
+  writer.byteOffset += bytes.length
+}
+/**
+ *
+ * @param {{bytes:Uint8Array}} writer
+ * @param {number[]} varint
+ * @param {Uint8Array} header
+ */
+const writeHeader = ({ bytes }, varint, header) => {
+  bytes.set(varint)
+  bytes.set(header, varint.length)
+}
+
+/**
+ * Attempts to estimate header size given number of roots it will have assuming
+ * they're CIDv1 in with sha256 digest. Optionally it takes number of bytes
+ * to be allocated for all the roots, in case different hashing or CID version
+ * is used.
+ *
+ * Note: Returned value is just an estimate which can be inaccurate where large
+ * number of CIDs is passed or if they are of various sizes.
+ *
+ * @param {number} count - Number of roots
+ * @param {number} [rootsByteLength] - Total byteLength of all roots
+ */
+export const estimateHeaderSize = (
+  count,
+  rootsByteLength = count * DEFAULT_CID_SIZE
+) =>
+  count > 0
+    ? headerEncodeSize(count, cidEncodeSize(Math.ceil(rootsByteLength / count)) * count)
+    : headerEncodeSize(0, 0)
+
+/**
+ *
+ * @param {number} count
+ * @param {number} rootsSize
+ * @returns
+ */
+const headerEncodeSize = (count, rootsSize) => {
+  const lengthSize = arrayLengthEncodeSize(count)
+  const headerSize = EMPTY_HEADER_SIZE + lengthSize + rootsSize
+  const varintSize = varint.encodingLength(headerSize)
+  return varintSize + headerSize
+}
+
+/**
+ *
+ * @param {number} length
+ * @returns
+ */
+const arrayLengthEncodeSize = length =>
+  length < 24
+    ? 1
+    : length < 256
+      ? 2
+      : length < 65536
+        ? 3
+        : CBOR.encode(length).length
+
+/**
+ * @param {number} byteLength
+ * @returns {number}
+ */
+const cidEncodeSize = byteLength =>
+  byteLength + arrayLengthEncodeSize(byteLength) + CID_TAG_SIZE
+
+/**
+ * @param {CID[]} cids
+ */
+const totalByteLength = (cids) => {
+  let total = 0
+  for (const cid of cids) {
+    total += cidEncodeSize(cid.bytes.byteLength)
+  }
+  return total
+}
+
+/**
+ * Creates synchronous CAR writer that can be used to encode blocks into a given
+ * buffer. Optionally you could pass `byteOffset` and `byteLength` to specify a
+ * range inside buffer to write into. If car file is going to have `roots` you
+ * need to either pass them under `options.roots` or provide `options.headerCapacity`
```

Comment:
`headerCapacity` here, `estimateHeaderCapacity` below and in test/test-buffer-writer.js seem to be leftovers from a previous iteration that had those names.